runtime/src/main/cpp/CyclicCollector.cpp

```diff
@@ -0,0 +1,460 @@
+  if (typeInfo != theArrayTypeInfo) {
+    for (int index = 0; index < typeInfo->objOffsetsCount_; index++) {
+      ObjHeader** location = reinterpret_cast<ObjHeader**>(
+          reinterpret_cast<uintptr_t>(obj) + typeInfo->objOffsets_[index]);
+      process(location);
+    }
+  } else {
+    ArrayHeader* array = obj->array();
+    for (int index = 0; index < array->count_; index++) {
+      process(ArrayAddressOfElementAt(array, index));
+    }
+  }
+}
+
+inline bool isAtomicReference(ObjHeader* obj) {
+  return (obj->type_info()->flags_ & TF_LEAK_DETECTOR_CANDIDATE) != 0;
+}
+
+#define CHECK_CALL(call, message) RuntimeCheck((call) == 0, message)
+
+class CyclicCollector {
+  pthread_mutex_t lock_;
+  pthread_cond_t cond_;
+  pthread_t gcThread_;
+
+  int currentAliveWorkers_;
+  int gcRunning_;
+  int mutatedAtomics_;
+  int pendingRelease_;
+  bool shallCollectGarbage_;
+  bool shallRunCollector_;
+  bool terminateCollector_;
+  int32_t currentTick_;
+  int32_t lastTick_;
+  int64_t lastTimestampUs_;
+  void* mainWorker_;
+  KStdUnorderedMap<ObjHeader*, int> sideRefCounts_;
+  KStdUnorderedSet<ObjHeader*> rootset_;
+  KStdUnorderedSet<ObjHeader*> toRelease_;
+
+ public:
+  CyclicCollector() {
+    CHECK_CALL(pthread_mutex_init(&lock_, nullptr), "Cannot init collector mutex")
+    CHECK_CALL(pthread_cond_init(&cond_, nullptr), "Cannot init collector condition")
+    CHECK_CALL(pthread_create(&gcThread_, nullptr, gcWorkerRoutine, this), "Cannot start collector thread")
+  }
+
+  ~CyclicCollector() {
+    {
+      Locker locker(&lock_);
+      terminateCollector_ = true;
+      shallRunCollector_ = true;
+      CHECK_CALL(pthread_cond_signal(&cond_), "Cannot signal collector")
+    }
+    // TODO: improve waiting for collector termination.
+    while (atomicGet(&terminateCollector_)) {}
+    releasePendingUnlocked(nullptr);
+    pthread_cond_destroy(&cond_);
+    pthread_mutex_destroy(&lock_);
+  }
+
+  static void* gcWorkerRoutine(void* argument) {
+    CyclicCollector* thiz = reinterpret_cast<CyclicCollector*>(argument);
+    thiz->gcProcessor();
+    return nullptr;
+  }
+
+  void gcProcessor() {
+     {
+       Locker locker(&lock_);
+       KStdDeque<ObjHeader*> toVisit;
+       KStdUnorderedSet<ObjHeader*> visited;
+       int restartCount = 0;
+       while (!terminateCollector_) {
+         CHECK_CALL(pthread_cond_wait(&cond_, &lock_), "Cannot wait collector condition")
+         if (!shallRunCollector_) continue;
+         atomicSet(&gcRunning_, 1);
+         restartCount = 0;
+        restart:
+         if (restartCount > 10) {
+           COLLECTOR_LOG("wait for some time to avoid GC trashing\n");
+           struct timeval tv;
+           struct timespec ts;
+           long long nsDelta = 1000LL * 1000LL * (restartCount - 10);
+           ts.tv_nsec = (tv.tv_usec * 1000LL + nsDelta) % 1000000000LL;
+           ts.tv_sec = (tv.tv_sec * 1000000000LL + tv.tv_usec * 1000000LL + nsDelta) / 1000000000LL ;
```

Comment:
> `tv.tv_usec * 1000000LL`

Why `1000000LL`?

Reply:
Indeed, that's microseconds, not milliseconds.